### PR TITLE
Made reactive services and repositories

### PR DESCRIPTION
**Pull request by ArSiu**

### CODE REVIEW

File: eduHub/src/main/kotlin/com/arsiu/eduhub/config/bpp/NotifyTriggerAnnotationHandlerBeanPostProcessor.kt [hunk 1]
```diff
@@ -62,14 +62,14 @@ class NotifyTriggerAnnotationHandlerBeanPostProcessor(
...
-                    (it.javaClass.typeParameters.contentEquals(method.javaClass.typeParameters))
+            (it.javaClass.typeParameters.contentEquals(method.javaClass.typeParameters))
         }?.findAnnotation<NotifyTrigger>()
     }
 
     @OptIn(DelicateCoroutinesApi::class)
     private fun notifyUsers(msg: String) {
         GlobalScope.launch {
-            userService.findAll()
+            userService.findAll().collectList().block()!!
```
Comment (ajax-gnatiuk-s) [inline eduHub/src/main/kotlin/com/arsiu/eduhub/config/bpp/NotifyTriggerAnnotationHandlerBeanPostProcessor.kt]: Suggestion:
```kotlin
userService.findAll()
    .parallel()
    .runOn(<Some-scheduler-here>)
    .filter { it.role == Role.STUDENT }
    .doOnNext { 
        emailServiceImpl.sendMail(EmailParameters(it.email, msg, msg)) 
    }
    .subscribe()
```  

So, each email could be sent in a separate thread. This code doesn't require usage of coroutinies.

File: eduHub/src/main/kotlin/com/arsiu/eduhub/repository/impl/AssignmentRepositoryImpl.kt [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+        findByIdCascade(entity.id)
+            .flatMap { existingEntity ->
+                if (existingEntity != entity) {
+                    val query = Query.query(Criteria.where("id").`is`(entity.id))
+
+                    val update = Update()
+                        .set("id", entity.id)
+                        .set("name", entity.name)
+
+                    reactiveMongoTemplate.upsert(query, update, Assignment::class.java).map { entity }
```
Comment (ajax-gnatiuk-s) [inline eduHub/src/main/kotlin/com/arsiu/eduhub/repository/impl/AssignmentRepositoryImpl.kt]:
```suggestion
                    reactiveMongoTemplate.upsert(query, update, Assignment::class.java).thenReturn(entity)
```

File: eduHub/src/main/kotlin/com/arsiu/eduhub/repository/impl/CourseRepositoryImpl.kt [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+
+                    deleteChapters.thenMany(updateChapters)
+                        .then(Mono.defer {
+                            val update = Update()
+                                .set("id", entity.id)
+                                .set("name", entity.name)
+                                .set("ownerId", entity.ownerId)
+                                .set("chapters", entity.chapters)
+                            reactiveMongoTemplate.upsert(query, update, Course::class.java).map { entity }
+
```
Comment (ajax-gnatiuk-s) [inline eduHub/src/main/kotlin/com/arsiu/eduhub/repository/impl/CourseRepositoryImpl.kt]:
```suggestion
```

Comment (ArSiu) [reply eduHub/src/main/kotlin/com/arsiu/eduhub/repository/impl/CourseRepositoryImpl.kt]: `+`

File: eduHub/src/main/kotlin/com/arsiu/eduhub/repository/impl/AssignmentRepositoryImpl.kt (hunk 2 above)
Comment (ArSiu) [reply eduHub/src/main/kotlin/com/arsiu/eduhub/repository/impl/AssignmentRepositoryImpl.kt]: `+`

File: eduHub/src/main/kotlin/com/arsiu/eduhub/config/bpp/NotifyTriggerAnnotationHandlerBeanPostProcessor.kt (hunk 1 above)
Comment (ArSiu) [reply eduHub/src/main/kotlin/com/arsiu/eduhub/config/bpp/NotifyTriggerAnnotationHandlerBeanPostProcessor.kt]: `+`

**Review by ajax-gnatiuk-s — APPROVED** [review]

**Pull request merged**